### PR TITLE
Add multiple-units support in search length criteria

### PR DESCRIPTION
**Pull request by solstice23**
Support multiple units in search length criteria

The following formats are now possible:

+ `7m27s`
+ `1h2m3s`
+ `7:27`
+ `1:2:3`

### CODE REVIEW

**Comment by bdach** [comment]
* Did you run the tests at all here? Several of them are failing.
* This parsing is very lax. I found several examples of wonky inputs that are accepted by this suggested parser in 5 minutes:
  * `0:4:1` (parses to 0:04:01, arguably shouldn't be allowed due to leading zeroes)
  * `0:3:` (parses to 0:03:00, one part is entirely skipped)
  * `0:` and `0:0:` (parse to zero)
  * `4s9m` (parses to 0:09:04, part order is inverted)
  * `2m2m2m2m2m` (parses to... 0:10:00)

**Review by bdach — CHANGES_REQUESTED** [review]
as above (at least the tests need to be fixed or adjusted to pass)

**Comment by solstice23** [comment]
Should we allow criteria with decimals? Like these:

+ `6.5m`
+ `6m15.5s`
+ `6.5m25.3s`

I think only the first two are valid, which solely have decimals in the last segment.

**Comment by bdach** [comment]
Probably but I wouldn't necessarily mind removing support for decimals entirely. They were never super intuitive to use with time which is generally base-60 or base-24.

**Comment by solstice23** [comment]
Now it should work fine.

These are valid:
+ `7m27s`
+ `1h2m3s`
+ `1h2m3.5s`
+ `7:27`
+ `1:2:3`
+ `1:02:03`
+ `6`
+ `6.5`
+ `6.5s`
+ `6.5m`

There are not valid:
+ `7.5m27s`
+ `7m27`
+ `7m7m7m`
+ `7m70s`
+ `5s6m`
+ `0:`
+ `0:3`

**Comment by bdach** [comment]
I would expect the test coverage to be expanded to cover the above cases, both positive and negative. And I still take issue with some of the accepted inputs (`1:2:3`?)

The code is also... not as I would write it. The regexes are a step forward but they're really complicated. Nested optional groups like `@"^\d+(:\d+(:\d+)?)?$"`, or the way that the `hms` ordering enforcement is done, are really way too difficult than they need to be. This could be vastly improved using capturing groups, for instance:

```
^((?<hours>\d+):)?(?<minutes>[0-5]\d):(?<seconds>[0-5]\d)$ // h:mm:ss format
^(?<minutes>[0-5]?\d):(?<seconds>[0-5]\d)$ // m:ss format
^((?<hours>\d+)h)?((?<minutes>[0-5]?\d)m)?((?<seconds>[0-5]?\d)s)?$ // XhXmXs format
```

The above examples are doing format enforcement *and* are simultaneously splitting parts out for better value extraction. I think they are far more understandable that what the current code is doing, but I suppose this is personal opinion.

**Comment by solstice23** [comment]
I suppose formats like `1:2:3` are acceptable because they are just simplified forms. `0:02:03` looks better but users' input without the prefix 0 should be allowed too. That increase the tolerance of the user input.

The nest regex has been simplified to `^\d+(:\d+){1,2}$`. Forgive me for not using the static format of regex because we accept inputs like `1h2m`, `1m2s`, and `1h2m3s` so there are multiple regexes which should be implemented if we separate each format like this. I think that will make the code redundant.

**Comment by bdach** [comment]
> there are multiple regexes which should be implemented if we separate each format like this

I think that the three I gave above are sufficient for any needs you may have. Maybe except for the fractional format, but as above, I say we drop it.

> Honestly I am not familiar with unit tests so I don't know how to add negative cases while the function tryUpdateLengthRange returns false.

You'd write negative tests by checking that the bad queries result in the `FilterCriteria` having `Length.HasFilter == false`.

**Comment by solstice23** [comment]
https://github.com/ppy/osu/pull/19275/files#diff-55ee49f7d5113ff54cf89f0567697cac68678297bf4fe046d0f3b1b8fb7658daR330-R332

3 regexes had been implemented instead of the previous splitting parsing method.

Additional, an extra test case `1h20s` had been added, which is valid.

File: osu.Game.Tests/NonVisual/Filtering/FilterQueryParserTest.cs [hunk 1]
```diff
@@ -120,24 +120,54 @@ public void TestApplyBPMQueries()
...
+            new object[] { "0:", new TimeSpan(), new TimeSpan(), false },
+            new object[] { ":0", new TimeSpan(), new TimeSpan(), false },
+            new object[] { "0:3:", new TimeSpan(), new TimeSpan(), false },
+            new object[] { "3:15.5", new TimeSpan(), new TimeSpan(), false },
         };
 
         [Test]
         [TestCaseSource(nameof(length_query_examples))]
-        public void TestApplyLengthQueries(string lengthQuery, TimeSpan expectedLength, TimeSpan scale)
+        public void TestApplyLengthQueries(string lengthQuery, TimeSpan expectedLength, TimeSpan scale, bool isValid)
```
Comment (bdach) [inline osu.Game.Tests/NonVisual/Filtering/FilterQueryParserTest.cs:155]: I don't generally much like when a single test has conditional branches. It complicates understanding, and also - as can be seen from `length_query_examples` - you have to pass garbage data to the negative test cases just to fit the method signature. I'd split the positive cases and the negative cases into separate test cases, like so:

```csharp
[Test]
[TestCaseSource(nameof(correct_length_query_examples))]
public void TestApplyLengthQueries(string lengthQuery, TimeSpan expectedLength, TimeSpan scale)
{
    // ...
}

[Test]
[TestCaseSource(nameof(incorrect_length_query_examples))]
public void TestInvalidLengthQueries(string query)
{
    // ...
}
```

File: osu.Game/Screens/Select/FilterQueryParser.cs [hunk 2]
```diff
@@ -312,11 +323,45 @@ private static bool tryUpdateCriteriaRange<T>(ref FilterCriteria.OptionalRange<T
...
         private static bool tryUpdateLengthRange(FilterCriteria criteria, Operator op, string val)
         {
-            if (!tryParseDoubleWithPoint(val.TrimEnd('m', 's', 'h'), out double length))
+            List<string> parts = new List<string>();
+            GroupCollection groups = null;
+
+            if (
+                tryMatchRegex(val, @"^((?<hours>\d+):)?(?<minutes>\d+):(?<seconds>\d+)$", ref groups) ||
+                tryMatchRegex(val, @"^((?<hours>\d+(\.\d+)?)h)?((?<minutes>\d+(\.\d+)?)m)?((?<seconds>\d+(\.\d+)?)s)?$", ref groups) ||
+                tryMatchRegex(val, @"^(?<seconds>\d+(\.\d+)?)$", ref groups)
```
Comment (bdach) [inline osu.Game/Screens/Select/FilterQueryParser.cs:332]: This all can't be in an `if` condition, it just reads wrong. Recommended pattern to use instead:

```csharp
Match? match = null;

match ??= tryMatchRegex(val, ...);
match ??= tryMatchRegex(val, ...);

if (match != null)
    // continue processing
```

File: osu.Game/Screens/Select/FilterQueryParser.cs [hunk 3]
```diff
@@ -312,11 +323,45 @@ private static bool tryUpdateCriteriaRange<T>(ref FilterCriteria.OptionalRange<T
...
+            )
+            {
+                if (groups["seconds"].Success)
+                    parts.Add(groups["seconds"].Value + "s");
+                if (groups["minutes"].Success)
+                    parts.Add(groups["minutes"].Value + "m");
+                if (groups["hours"].Success)
+                    parts.Add(groups["hours"].Value + "h");
```
Comment (bdach) [inline osu.Game/Screens/Select/FilterQueryParser.cs:340]: I'm not sure why this is assembling strings anew, even though the input has just been parsed into recognisable groups. This shouldn't be required. The input is already recognised, use the matches directly.

Comment (solstice23) [reply osu.Game/Screens/Select/FilterQueryParser.cs:340]: Reassembling them into an array makes them easier to check. Each of them is conditional (in other words every part can exist or not). And we should check if they satisfy these rules (assume they are ordered in any subsequence of `s, m, h`):
+ Each part are numbers (int or decimal) (already checked in the regexes)
+ All of them should be less than 60 except the last part
+ All of them shouldn't contain a decimal dot except the first part
+ Determine the minimum scale

If we assemble them into a sequence, all these checks are easy to do. But if we use the match directly, it should take us a lot of codes to check them.

Don't know if there is a better way exists, so I suppose to keep the reassembling method for now.

File: osu.Game/Screens/Select/FilterQueryParser.cs [hunk 4]
```diff
@@ -312,11 +322,43 @@ private static bool tryUpdateCriteriaRange<T>(ref FilterCriteria.OptionalRange<T
...
+            match ??= tryMatchRegex(val, @"^((?<hours>\d+):)?(?<minutes>\d+):(?<seconds>\d+)$");
+            match ??= tryMatchRegex(val, @"^((?<hours>\d+(\.\d+)?)h)?((?<minutes>\d+(\.\d+)?)m)?((?<seconds>\d+(\.\d+)?)s)?$");
+            match ??= tryMatchRegex(val, @"^(?<seconds>\d+(\.\d+)?)$");
+
+            if (match == null)
                 return false;
 
-            int scale = getLengthScale(val);
-            return tryUpdateCriteriaRange(ref criteria.Length, op, length * scale, scale / 2.0);
+            if (match["seconds"].Success)
```
Comment (solstice23) [inline osu.Game/Screens/Select/FilterQueryParser.cs:335]: Reassembling is necessary unless there is a better way to do all checks without assembling these match results to an ordered sequence.

The checks which rely on the array:
- All of them should be less than 60 except the last part
- All of them shouldn't contain a decimal dot except the first part
- Determine the minimum scale

**Review by peppy — APPROVED** [review]

**Pull request merged**